test/disjointCoarseMallocPool.cpp

```diff
@@ -0,0 +1,610 @@
+        }
+
+        for (int j = 0; j < 8; j++) {
+            umf_result_t res = umfPoolFree(pool, t[j]);
+            ASSERT_EQ(res, UMF_RESULT_SUCCESS);
+        }
+    }
+
+    umfPoolDestroy(pool);
+    umfMemoryProviderDestroy(coarse_memory_provider);
+    umfMemoryProviderDestroy(malloc_memory_provider);
+}
+
+struct alloc_ptr_size {
+    void *ptr;
+    size_t size;
+
+    bool operator<(const alloc_ptr_size &other) const {
+        if (ptr == other.ptr) {
+            return size < other.size;
+        }
+        return ptr < other.ptr;
+    }
+};
+
+TEST_F(test, disjointCoarseMMapPool_random) {
+
+    const size_t KB = 1024;
+    const size_t MB = 1024 * KB;
+    const size_t init_buffer_size = 200 * MB;
+
+    // Preallocate some memory
+    void *buf = malloc(init_buffer_size);
+    memset(buf, 0, init_buffer_size);
+    ASSERT_NE(buf, nullptr);
+
+    const unsigned char alloc_check_val = 11;
+    coarse_memory_provider_params_t coarse_memory_provider_params = {
+        NULL,  // upstream_memory_provider
+        false, // immediate_init - this is used only with the upstream provider
+        buf,   // ptr to preallocated memory
+        init_buffer_size,
+        true,  // trace
+        false, // WA_do_not_free_upstream
+    };
+
+    umf_memory_provider_handle_t coarse_memory_provider;
+    umfMemoryProviderCreate(umfCoarseMemoryProviderOps(),
+                            &coarse_memory_provider_params,
+                            &coarse_memory_provider);
+    ASSERT_NE(coarse_memory_provider, nullptr);
+
+    umf_disjoint_pool_params_t disjoint_memory_pool_params = {};
+    disjoint_memory_pool_params.SlabMinSize = 1024;
+    disjoint_memory_pool_params.MaxPoolableSize = 1024;
+    disjoint_memory_pool_params.Capacity = 2;
+    disjoint_memory_pool_params.MinBucketSize = 16;
+    disjoint_memory_pool_params.PoolTrace = 1;
+
+    umf_memory_pool_handle_t pool;
+    umfPoolCreate(umfDisjointPoolOps(), coarse_memory_provider,
+                  &disjoint_memory_pool_params, UMF_POOL_CREATE_FLAG_NONE,
+                  &pool);
+    ASSERT_NE(pool, nullptr);
+
+    // set constant seed so each test run will have the same scenario
+    uint32_t seed = 1234;
+    std::mt19937 mt(seed);
+
+    // different sizes to alloc
+    std::vector<size_t> sizes = {15,        49,     588,      1025,
+                                 2 * KB,    5 * KB, 160 * KB, 511 * KB,
+                                 1000 * KB, MB,     3 * MB,   7 * MB};
+    std::uniform_int_distribution<int> sizes_dist(0, (int)(sizes.size() - 1));
+
+    // each alloc would be done few times
+    std::vector<size_t> counts = {1, 3, 4, 8, 9, 11};
+    std::uniform_int_distribution<int> counts_dist(0, (int)(counts.size() - 1));
+
+    // action to take will be random
+    // alloc = <0, .5), free = <.5, 1)
+    std::uniform_real_distribution<float> actions_dist(0, 1);
+
+    std::set<alloc_ptr_size> allocs;
+    for (size_t i = 0; i < 100; i++) {
```

Comment:
This whole loop could be simplified (and made more readable) if you'd implement the logic for choosing size/action, etc. in a separate function/class:

```
allocation_state state; // holds std::set<alloc_ptr_size>
for (size_t i = 0; i < 100; i++) {
   auto action = generator.get_action();
   if (action.is_alloc()) {
       state.do_alloc(action.count, ...);
   } else {
       state.do_free(action.acount, ...);
   }
}
```

Reply:
Also, this way you could reuse this for other tests (even multithreaded perhaps)